### ISSUE TITLE
Clarification regarding best practice for pad_token

### ISSUE THREAD

**Issue by hanyin88** [facebookresearch/llama-recipes#298]
### System Info

Build from souce code.

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

I have noticed several discrepancies in how to set up pad_token from official documentations, and would like to seek some kind clarification. 

From [Hugginface](https://huggingface.co/docs/transformers/main/model_doc/llama2), it was mentioned that "The original model uses pad_id = -1 which means that there is no padding token. We can’t have the same logic, make sure to add a padding token using tokenizer.add_special_tokens({"pad_token":"<pad>"}) and resize the token embedding accordingly. You should also set the model.config.pad_token_id. The embed_tokens layer of the model is initialized with self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.config.padding_idx), which makes sure that encoding the padding token will output zeros, so passing it when initializing is recommended."

From the[llama-recipes/inference.md](https://github.com/facebookresearch/llama-recipes/blob/3b77a270a87b7a8e4084402bd94017d8b88e5458/docs/inference.md), it was mentioned that "Note Currently pad token by default in [HuggingFace Tokenizer is None](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/tokenization_llama.py#L110). We add the padding token as a special token to the tokenizer, which in this case requires to resize the token_embeddings as shown below:

`tokenizer.add_special_tokens(
      {
            "pad_token": "<PAD>",
        }
    )`
`model.resize_token_embeddings(model.config.vocab_size + 1)`

Padding would be required for batch inference. In this this [example](https://github.com/facebookresearch/llama-recipes/blob/3b77a270a87b7a8e4084402bd94017d8b88e5458/examples/inference.py), batch size = 1 so essentially padding is not required. However,We added the code pointer as an example in case of batch inference." Therefore in this approach we changed pad_token following by embedding resize, but did not adjust  pad_token_id.

However, in the actual [inference.py](https://github.com/facebookresearch/llama-recipes/blob/3b77a270a87b7a8e4084402bd94017d8b88e5458/examples/inference.py#L75) and [fiinetuning.py](https://github.com/facebookresearch/llama-recipes/blob/3b77a270a87b7a8e4084402bd94017d8b88e5458/src/llama_recipes/finetuning.py#L119), what we did is:
` tokenizer.pad_token = tokenizer.eos_token`
This approach was not followed by any embedding resizing or change in pad_token_id. 

Overall it seems to me the third approach (setting tokenizer.pad_token = tokenizer.eos_token) is the easiest method, but just wish to confirm there won't be downstream issues. 

Many thanks!

### Error logs

N/A

### Expected behavior

N/A

**Comment by HamidShojanazeri** [comment]
@hanyin88, @qianjyM  sorry for the late reply currently, we support both packing and padding dataset mode, packing  basically avoid padding in the fine-tuning phase for more compute efficiency. The `set_tokenizer_params` is not being used during fine-tuning (legacy code need to be removed) so still we do set ` tokenizer.pad_token = tokenizer.eos_token`.

**Comment by ShengYun-Peng** [comment]
Continuing the above discussion, I believe the [inference.py](https://github.com/meta-llama/llama-recipes/blob/main/recipes/inference/local_inference/inference.py) is missing `tokenizer.padding_side = "left"`. Although current inference.py only supports a single prompt and no padding/left padding is fine, it's incorrect if we scale up and do a batch inference. The first image below shows that w/o left padding, the generated text is weird except the first prompt. It's because the first prompt is the longest and does not need padding. I explicitly expose the special tokens to highlight the paddings between the prompt and the generated output. However, if we switch to left padding, we get the correct output as shown in the second image below. 

Image 1: right padding
<img width="1284" alt="Screenshot 2024-04-15 at 11 39 26 AM" src="https://github.com/meta-llama/llama-recipes/assets/84164548/7b3b564d-58e9-424c-b754-381c11142834">

Image 2: left padding
<img width="1292" alt="Screenshot 2024-04-15 at 11 42 55 AM" src="https://github.com/meta-llama/llama-recipes/assets/84164548/6e402d75-a262-49c0-8605-b232dec7e10f">

**Comment by HamidShojanazeri** [comment]
thanks @ShengYun-Peng would be insterested to send a PR?
